code/modules/projectiles/projectile.dm

```diff
@@ -674,268 +735,412 @@
+		if (movement_vector.pixel_x)
+			x_to_border = movement_vector.pixel_x > 0 ? (ICON_SIZE_X - pixel_x_actual) / movement_vector.pixel_x : pixel_x_actual / -movement_vector.pixel_x
+		if (movement_vector.pixel_y)
+			y_to_border = movement_vector.pixel_y > 0 ? (ICON_SIZE_Y - pixel_y_actual) / movement_vector.pixel_y : pixel_y_actual / -movement_vector.pixel_y
+
+		// Figure out how much we need to travel to hit the border
+		var/distance_to_border = !isnull(x_to_border) ? (!isnull(y_to_border) ? min(x_to_border, y_to_border) : x_to_border) : (!isnull(y_to_border) ? y_to_border : -1)
+		// Something went extremely wrong
+		if (distance_to_border < 0)
+			stack_trace("WARNING: Projectile had an empty movement vector and tried to process")
 			qdel(src)
 			return
-		if (cur_turf == loc)
-			continue
-		if (cur_turf.z == loc.z)
-			step_towards(src, cur_turf)
-			hitscan_last = loc
-			SEND_SIGNAL(src, COMSIG_PROJECTILE_PIXEL_STEP)
-			continue
-		var/old = loc
-		before_z_change(loc, cur_turf)
-		trajectory_ignore_forcemove = TRUE
-		forceMove(cur_turf)
-		trajectory_ignore_forcemove = FALSE
-		after_z_change(old, loc)
-		if(!hitscanning)
-			pixel_x = trajectory.return_px()
-			pixel_y = trajectory.return_py()
-		forcemoved = TRUE
-		hitscan_last = loc
-		SEND_SIGNAL(src, COMSIG_PROJECTILE_PIXEL_STEP)
-	if(QDELETED(src)) //deleted on last move
-		return
-	if(!hitscanning && !forcemoved)
-		pixel_x = trajectory.return_px() - trajectory.mpx * trajectory_multiplier * SSprojectiles.global_iterations_per_move
-		pixel_y = trajectory.return_py() - trajectory.mpy * trajectory_multiplier * SSprojectiles.global_iterations_per_move
-		animate(src, pixel_x = trajectory.return_px(), pixel_y = trajectory.return_py(), time = 1, flags = ANIMATION_END_NOW)
-	Range()
 
-/obj/projectile/proc/process_homing() //may need speeding up in the future performance wise.
+		var/distance_to_move = min(distance_to_border, pixels_to_move)
+		// For homing we cap the maximum distance to move every loop
+		if (homing && distance_to_move > SSprojectiles.pixels_per_decisecond)
+			distance_to_move = SSprojectiles.pixels_per_decisecond
+
+		// Figure out if we move to the next turf and if so, what its positioning relatively to us is
+		var/x_shift = SIGN(movement_vector.pixel_x) * (!isnull(x_to_border) && distance_to_move >= x_to_border)
+		var/y_shift = SIGN(movement_vector.pixel_y) * (!isnull(y_to_border) && distance_to_move >= y_to_border)
+		var/turf/new_turf = (x_shift || y_shift) ? locate(x + x_shift, y + y_shift, z) : loc
+		// Calculate where in the turf we will be when we cross the edge.
+		// This is a projectile variable because its also used in hit VFX
+		entry_x = pixel_x + movement_vector.pixel_x * distance_to_move - x_shift * ICON_SIZE_X
+		entry_y = pixel_y + movement_vector.pixel_y * distance_to_move - y_shift * ICON_SIZE_Y
+		var/delete_distance
+
+		if (x_shift || y_shift)
+			// We've hit an invalid turf, end of a z level or smth went wrong
```

Comment:
new_turf does not need to be in the outer scope
```suggestion
		// This is a projectile variable because its also used in hit VFX
		entry_x = pixel_x + movement_vector.pixel_x * distance_to_move - x_shift * ICON_SIZE_X
		entry_y = pixel_y + movement_vector.pixel_y * distance_to_move - y_shift * ICON_SIZE_Y
		var/delete_distance

		if (moving_turfs)
			// Calculate where in the turf we will be when we cross the edge.
			var/turf/new_turf = locate(x + x_shift, y + y_shift, z)
			// We've hit an invalid turf, end of a z level or smth went wrong
```